import "yt/yt_proto/yt/formats/extension.proto";

package NYT.NTesting;

message TUrlRow {
    optional string Host = 1     [(NYT.column_name) = "Host"];
    optional string Path = 2     [(NYT.column_name) = "Path"];
    optional sint32 HttpCode = 3 [(NYT.column_name) = "HttpCode"];
}

message TGoodUrl {
    optional string Url = 1     [(NYT.column_name) = "Url"];
}

message THostRow {
    optional string Host = 1          [(NYT.column_name) = "Host"];
    optional sint32 HttpCodeTotal = 2 [(NYT.column_name) = "HttpCodeTotal"];
}

message TRowVer1 {
    optional string String_1 = 1;
    optional uint32 Uint32_2 = 2;
}

message TRowVer2 {
    optional string String_1 = 1;
    optional uint32 Uint32_2 = 2;
    optional fixed64 Fixed64_3 = 3;
}

message TRowFieldSerializationOption {
    optional TUrlRow UrlRow_1 = 1 [(NYT.flags) = SERIALIZATION_YT];
    optional TUrlRow UrlRow_2 = 2;
}

message TRowMessageSerializationOption {
    option (NYT.default_field_flags) = SERIALIZATION_YT;
    optional TUrlRow UrlRow_1 = 1;
    optional TUrlRow UrlRow_2 = 2;
}

message TRowMixedSerializationOptions {
    option (NYT.default_field_flags) = SERIALIZATION_YT;
    optional TUrlRow UrlRow_1 = 1;
    optional TUrlRow UrlRow_2 = 2 [(NYT.flags) = SERIALIZATION_PROTOBUF];
}

message TRowSerializedRepeatedFields {
    option (NYT.default_field_flags) = SERIALIZATION_YT;
    repeated int64 Ints = 1;
    repeated int64 PackedInts = 2 [packed=true];
    repeated TUrlRow UrlRows = 3;
}

message TUrlRowWithColumnNames {
    optional string Host = 1     [(NYT.column_name) = "Host_ColumnName", (NYT.key_column_name) = "Host_KeyColumnName"];
    optional string Path = 2     [(NYT.key_column_name) = "Path_KeyColumnName"];
    optional sint32 HttpCode = 3;
}

message TRowMixedSerializationOptions_ColumnNames {
    option (NYT.default_field_flags) = SERIALIZATION_YT;
    optional TUrlRowWithColumnNames UrlRow_1 = 1;
    optional TUrlRowWithColumnNames UrlRow_2 = 2 [(NYT.flags) = SERIALIZATION_PROTOBUF];
}

message TBadProtobufSerializedRow {
    // Repeated fields are not supported in PROTOBUF serialization mode.
    repeated int64 Ints = 1;
}

message TRowWithTypeOptions
{
    enum Color
    {
        WHITE = 0;
        BLUE = 1;
        RED = -1;
    }

    message Embedded
    {
        optional Color ColorIntField = 1 [(NYT.flags) = ENUM_INT];
        optional Color ColorStringField = 2 [(NYT.flags) = ENUM_STRING];
        optional bytes AnyField = 3 [(NYT.flags) = ANY];
    }

    optional Color ColorStringField = 2 [(NYT.flags) = ENUM_STRING];
    optional bytes OtherColumnsField = 4 [(NYT.flags) = OTHER_COLUMNS];
    optional Color ColorIntField = 1 [(NYT.flags) = ENUM_INT];
    optional Embedded EmbeddedField = 5 [(NYT.flags) = SERIALIZATION_YT];
    repeated Color RepeatedEnumIntField = 6 [(NYT.flags) = SERIALIZATION_YT, (NYT.flags) = ENUM_INT];
    optional bytes AnyField = 3 [(NYT.flags) = ANY];
}

message TGrepperRecord
{
    optional string Key = 1;
    optional bytes Other = 2 [(NYT.flags) = OTHER_COLUMNS];
}
